src/sanitize.c

```diff
@@ -0,0 +1,195 @@
+	sanitize_handlers[IORING_OP_MKDIRAT] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_SYMLINKAT] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_LINKAT] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_MSG_RING] = sanitize_sqe_addr_and_add3;
+	sanitize_handlers[IORING_OP_FSETXATTR] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_SETXATTR] = sanitize_sqe_addr_and_add3;
+	sanitize_handlers[IORING_OP_FGETXATTR] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_GETXATTR] = sanitize_sqe_addr_and_add3;
+	sanitize_handlers[IORING_OP_SOCKET] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_URING_CMD] = sanitize_sqe_optval;
+	sanitize_handlers[IORING_OP_SEND_ZC] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_SENDMSG_ZC] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_READ_MULTISHOT] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_WAITID] = sanitize_sqe_addr_and_add2;
+	sanitize_handlers[IORING_OP_FUTEX_WAIT] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_FUTEX_WAKE] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_FUTEX_WAITV] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_FIXED_FD_INSTALL] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_FTRUNCATE] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_BIND] = sanitize_sqe_addr;
+	sanitize_handlers[IORING_OP_LISTEN] = sanitize_sqe_addr;
+	sanitize_handlers_initialized = true;
+}
+
+void liburing_sanitize_ring(struct io_uring *ring)
+{
+	struct io_uring_sq *sq = &ring->sq;
+	struct io_uring_sqe *sqe;
+	unsigned int head;
+	int shift = 0;
+
+	initialize_sanitize_handlers();
+
+	if (ring->flags & IORING_SETUP_SQE128)
+		shift = 1;
+	if (!(ring->flags & IORING_SETUP_SQPOLL))
+		head = *sq->khead;
+	else
+		head = io_uring_smp_load_acquire(sq->khead);
+
+	while (head != sq->sqe_tail) {
+		sqe = &sq->sqes[(head & sq->ring_mask) << shift];
+		if (sqe->opcode < IORING_OP_LAST)
+			sanitize_handlers[sqe->opcode](sqe);
+		head++;
+	}
+}
+
+void liburing_sanitize_address(const void *addr)
+{
+	if (__asan_address_is_poisoned(addr) != 0) {
+		__asan_describe_address((void *)addr);
+		exit(1);
+	}
+}
+
+void liburing_sanitize_region(const void *addr, unsigned int len)
+{
+	if (__asan_region_is_poisoned((void *)addr, len) != 0) {
+		__asan_describe_address((void *)addr);
+		exit(1);
+	}
+}
+
+void liburing_sanitize_iovecs(const struct iovec *iovecs, unsigned nr)
+{
+	unsigned i;
+
+	if (__asan_address_is_poisoned((void *)iovecs) != 0) {
+		__asan_describe_address((void *)iovecs);
+		exit(1);
+	}
+
+	for (i = 0; i < nr; i++) {
+		if (__asan_region_is_poisoned((void *)iovecs[i].iov_base, iovecs[i].iov_len) != 0) {
+			__asan_describe_address((void *)iovecs[i].iov_base);
+			exit(1);
+		}
+	}
+}
+#else
+void liburing_sanitize_ring(struct io_uring *ring)
+{
+}
+
+void liburing_sanitize_address(const void *addr)
+{
+}
+
+void liburing_sanitize_region(const void *addr, unsigned int len)
+{
+}
+
+void liburing_sanitize_iovecs(const struct iovec *iovecs, unsigned nr)
+{
+}
+#endif
```

Comment:
These functions leave many calls in the lib even when **the sanitizer is not used**:
```
ammarfaizi2@integral2:~/lbq$ objdump -d src/liburing.so.2.8 | grep liburing_sanitize
    3629:	e8 c2 18 00 00       	call   4ef0 <liburing_sanitize_ring>
    36b3:	e8 38 18 00 00       	call   4ef0 <liburing_sanitize_ring>
    375d:	e8 8e 17 00 00       	call   4ef0 <liburing_sanitize_ring>
    37ee:	e8 fd 16 00 00       	call   4ef0 <liburing_sanitize_ring>
    3893:	e8 58 16 00 00       	call   4ef0 <liburing_sanitize_ring>
    3f2d:	e8 ee 0f 00 00       	call   4f20 <liburing_sanitize_iovecs>
    3f5b:	e8 a0 0f 00 00       	call   4f00 <liburing_sanitize_address>
    3fcd:	e8 4e 0f 00 00       	call   4f20 <liburing_sanitize_iovecs>
    3fec:	e8 0f 0f 00 00       	call   4f00 <liburing_sanitize_address>
    4061:	e8 9a 0e 00 00       	call   4f00 <liburing_sanitize_address>
    40c5:	e8 56 0e 00 00       	call   4f20 <liburing_sanitize_iovecs>
    40cd:	e8 2e 0e 00 00       	call   4f00 <liburing_sanitize_address>
    411a:	e8 e1 0d 00 00       	call   4f00 <liburing_sanitize_address>
    418f:	e8 6c 0d 00 00       	call   4f00 <liburing_sanitize_address>
    4197:	e8 64 0d 00 00       	call   4f00 <liburing_sanitize_address>
    41c6:	e8 35 0d 00 00       	call   4f00 <liburing_sanitize_address>
    422d:	e8 ce 0c 00 00       	call   4f00 <liburing_sanitize_address>
    4249:	e8 b2 0c 00 00       	call   4f00 <liburing_sanitize_address>
    42cc:	e8 2f 0c 00 00       	call   4f00 <liburing_sanitize_address>
    4391:	e8 6a 0b 00 00       	call   4f00 <liburing_sanitize_address>
    439b:	e8 60 0b 00 00       	call   4f00 <liburing_sanitize_address>
    43bb:	e8 40 0b 00 00       	call   4f00 <liburing_sanitize_address>
    446c:	e8 8f 0a 00 00       	call   4f00 <liburing_sanitize_address>
    4474:	e8 87 0a 00 00       	call   4f00 <liburing_sanitize_address>
    450a:	e8 f1 09 00 00       	call   4f00 <liburing_sanitize_address>
    456a:	e8 91 09 00 00       	call   4f00 <liburing_sanitize_address>
    45ba:	e8 41 09 00 00       	call   4f00 <liburing_sanitize_address>
    461a:	e8 e1 08 00 00       	call   4f00 <liburing_sanitize_address>
    4673:	e8 88 08 00 00       	call   4f00 <liburing_sanitize_address>
    46ba:	e8 41 08 00 00       	call   4f00 <liburing_sanitize_address>
    4711:	e8 ea 07 00 00       	call   4f00 <liburing_sanitize_address>
    4763:	e8 98 07 00 00       	call   4f00 <liburing_sanitize_address>
    47aa:	e8 51 07 00 00       	call   4f00 <liburing_sanitize_address>
    480d:	e8 ee 06 00 00       	call   4f00 <liburing_sanitize_address>
    486a:	e8 91 06 00 00       	call   4f00 <liburing_sanitize_address>
    48c4:	e8 37 06 00 00       	call   4f00 <liburing_sanitize_address>
    4946:	e8 b5 05 00 00       	call   4f00 <liburing_sanitize_address>
    49fe:	e8 fd 04 00 00       	call   4f00 <liburing_sanitize_address>
    4ae8:	e8 13 04 00 00       	call   4f00 <liburing_sanitize_address>
    4b5f:	e8 9c 03 00 00       	call   4f00 <liburing_sanitize_address>
    4bb8:	e8 43 03 00 00       	call   4f00 <liburing_sanitize_address>
    4bdd:	e8 1e 03 00 00       	call   4f00 <liburing_sanitize_address>
    4c54:	e8 a7 02 00 00       	call   4f00 <liburing_sanitize_address>
    4cce:	e8 2d 02 00 00       	call   4f00 <liburing_sanitize_address>
    4d24:	e8 d7 01 00 00       	call   4f00 <liburing_sanitize_address>
    4d84:	e8 77 01 00 00       	call   4f00 <liburing_sanitize_address>
    4de4:	e8 17 01 00 00       	call   4f00 <liburing_sanitize_address>
0000000000004ef0 <liburing_sanitize_ring>:
0000000000004f00 <liburing_sanitize_address>:
0000000000004f10 <liburing_sanitize_region>:
0000000000004f20 <liburing_sanitize_iovecs>:
```

I suggest making them empty functions with `static inline` in the header if the sanitizer is not used. That way the compiler can omit them.

The header would look like this:
```c
/* SPDX-License-Identifier: MIT */
#ifndef LIBURING_SANITIZE_H
#define LIBURING_SANITIZE_H

struct io_uring;
struct iovec;

#if defined(CONFIG_USE_SANITIZER)
void liburing_sanitize_ring(struct io_uring *ring);
void liburing_sanitize_address(const void *addr);
void liburing_sanitize_region(const void *addr, unsigned int len);
void liburing_sanitize_iovecs(const struct iovec *iovecs, unsigned nr);
#else
static inline void liburing_sanitize_ring(struct io_uring *ring) {}
static inline void liburing_sanitize_address(const void *addr) {}
static inline void liburing_sanitize_region(const void *addr, unsigned int len) {}
static inline void liburing_sanitize_iovecs(const struct iovec *iovecs, unsigned nr) {}
#endif

#endif
```

Reply:
You probably also don't have to compile sanitizer.c when CONFIG_USE_SANITIZER is not enabled.

Reply:
Indeed, these need to be static inline stubs in the header. Function calls aren't free...

Reply:
Nice find. Changed.